contracts/src/StakeTable.sol

```diff
@@ -52,493 +102,352 @@ contract StakeTable is AbstractStakeTable, Ownable, InitializedAt {
 
-        emit Registered(msg.sender, registrationEpoch, amount);
+        emit ValidatorExit(msg.sender);
     }
 
-    /// @notice Deposit more stakes to registered keys
-    /// @dev TODO this implementation will be revisited later. See
-    /// https://github.com/EspressoSystems/espresso-sequencer/issues/806
-    /// @dev TODO modify this according to the current spec
-    /// @param amount The amount to deposit
-    /// @return (newBalance, effectiveEpoch) the new balance effective at a future epoch
-    function deposit(uint256 amount) external virtual override returns (uint256, uint64) {
-        Node memory node = nodes[msg.sender];
-
-        if (node.account == address(0)) {
-            revert NodeNotRegistered();
-        }
-
-        if (node.account != msg.sender) {
-            revert Unauthenticated();
-        }
-
-        // A node cannot deposit more tokens while it waiting to register.
-        uint64 _currentEpoch = currentEpoch();
-        if (_currentEpoch <= node.registerEpoch) {
-            revert PrematureDeposit();
-        }
+    /// @notice Delegate to a validator
+    /// @param validator The validator to delegate to
+    /// @param amount The amount to delegate
+    function delegate(address validator, uint256 amount) external virtual {
+        ensureValidatorActive(validator);
 
-        // A node cannot deposit more tokens if an exit request is in progress.
-        if (node.exitEpoch != 0) {
-            revert ExitRequestInProgress();
+        uint256 allowance = token.allowance(msg.sender, address(this));
+        if (allowance < amount) {
+            revert InsufficientAllowance(allowance, amount);
         }
 
-        nodes[msg.sender].balance += amount;
-        SafeTransferLib.safeTransferFrom(ERC20(tokenAddress), msg.sender, address(this), amount);
-
-        emit Deposit(msg.sender, uint256(amount));
+        validators[validator].delegatedAmount += amount;
+        delegations[validator][msg.sender] += amount;
 
-        uint64 effectiveEpoch = _currentEpoch + 1;
+        SafeTransferLib.safeTransferFrom(token, msg.sender, address(this), amount);
 
-        return (nodes[msg.sender].balance, effectiveEpoch);
+        emit Delegated(msg.sender, validator, amount);
     }
 
-    /// @notice Request to exit from the stake table, not immediately withdrawable!
-    ///
-    /// @dev TODO modify this according to the current spec
-    function requestExit() external virtual override {
-        Node memory node = nodes[msg.sender];
+    /// @notice Undelegate from a validator
+    /// @param validator The validator to undelegate from
+    /// @param amount The amount to undelegate
+    function undelegate(address validator, uint256 amount) external virtual {
+        ensureValidatorActive(validator);
 
-        if (node.account == address(0)) {
-            revert NodeNotRegistered();
+        if (validators[msg.sender].status == ValidatorStatus.Exited) {
```

Comment:
hi hi, why can't a user undelegate? when a validator exits - are the delegated amounts already sent to the delegatos? if not, they should be able to undelegate at any point? based on my current understanding of the flow

Reply:
The validator exit will undelegate all the funds of the delegators.

Reply:
ok i understand that now, thanks!